### PR TITLE
 dhall-json: Nesting: Support empty alternatives as contents

### LINKED ISSUE #1201: Empty alternative does not trigger JSON name preservation
Possibly related to https://github.com/dhall-lang/dhall-haskell/pull/1080, https://github.com/dhall-lang/dhall-haskell/issues/1145

As per the `dhall-json` tests `nesting{0,1}`, the use of a record with fields `field`, `nesting`, and `contents` is supposed to preserve the name of the union alternative alongside the value, either Inline or Nested. However the behaviour does not trigger when the union alternative is empty.

For example, with this input:

```dhall
let Example = < Left : { foo : Natural } | Middle | Right : { bar : Bool } >

let Nesting = < Inline | Nested : Text >

in  { field    = "name"
    , nesting  = Nesting.Inline
    , contents = Example.Middle
    }
```

I expect this output:

```json
{ "name": "Middle" }
```

But instead get:

```json
{ "field": "name", "contents": "Middle", "nesting": "Inline" }
```

I have added a test case for this in my fork but I'm not familiar enough with Haskell to implement a fix: https://github.com/dhall-lang/dhall-haskell/compare/master...wezm:json-nesting-empty-alternative

### PR DESCRIPTION
**Pull request by sjakobi**
Closes #1201.

### CODE REVIEW

File: dhall-json/src/Dhall/JSON.hs [hunk 1]
```diff
@@ -319,12 +322,7 @@ dhallToJSON e0 = loop (Core.alphaNormalize (Core.normalize e0))
...
                 [   (   "contents"
-                    ,   Core.App
-                            (Core.Field
-                                _
-                                alternativeName
-                            )
-                            contents
+                    ,   contents
                     )
                  ,  (   "field"
```
Comment (sjakobi) [inline dhall-json/src/Dhall/JSON.hs]: The additional indentation here confused my for so long, I ended up opening [a GHC issue](https://gitlab.haskell.org/ghc/ghc/issues/17028)! ;)

File: dhall-json/src/Dhall/JSON.hs [hunk 2]
```diff
@@ -374,19 +370,35 @@ dhallToJSON e0 = loop (Core.alphaNormalize (Core.normalize e0))
...
-                                field
-                                (Core.TextLit
-                                    (Core.Chunks
-                                        []
-                                        alternativeName
-                                    )
-                                )
-                                contents
+                 ] | isInlineNesting nesting
+                   , Just (alternativeName, Just (Core.RecordLit kvs)) <- getContents contents -> do
```
Comment (sjakobi) [inline dhall-json/src/Dhall/JSON.hs]: I wonder whether we should report an error in this case:
```suggestion
                   , Just (alternativeName, Just notARecordLit) <- getContents contents -> do
```

Non-record alternatives are supported with `Nested`, so users might also try them with `Inline`.

File: dhall-json/src/Dhall/JSON.hs (hunk 1 above)
Comment (Gabriel439) [reply dhall-json/src/Dhall/JSON.hs]: Yeah, the indentation there bugged me, too 🙂

File: dhall-json/src/Dhall/JSON.hs (hunk 2 above)
Comment (Gabriel439) [reply dhall-json/src/Dhall/JSON.hs]: @sjakobi: Yeah, we probably should report an error here.  It's up to you if you want to include that in this pull request or not

Comment (sjakobi) [reply dhall-json/src/Dhall/JSON.hs]: I'll add the error in a follow-up PR.

**Comment by wezm** [comment]
Thank you!

**Pull request merged**